load-balancer/least-requests/src/test/java/io/smallrye/stork/loadbalancer/leastresponsetime/LeastRequestsLoadBalancerTest.java

```diff
@@ -0,0 +1,171 @@
+        // Now select 1 as they have the same amount of inflight
+        instance = selectInstance(service);
+        assertThat(asString(instance)).isEqualTo(FST_SRVC_1);
+        assertThat(instance.gatherStatistics()).isTrue();
+    }
+
+    @Test
+    public void shouldSelectLessLoadedWhenBothLoaded() {
+        Service service = stork.getService("first-service");
+
+        ServiceInstance service1, service2;
+
+        // First selection returns the first service
+        ServiceInstance instance = selectInstance(service);
+        assertThat(asString(instance)).isEqualTo(FST_SRVC_1);
+        assertThat(instance.gatherStatistics()).isTrue();
+        service1 = instance;
+
+        // Second selection returns the second service as there is an inflight-call in the first one
+        instance = selectInstance(service);
+        assertThat(asString(instance)).isEqualTo(FST_SRVC_2);
+        assertThat(instance.gatherStatistics()).isTrue();
+        service2 = instance;
+
+        ServiceInstance last = service2;
+        for (int i = 0; i < 100; i++) {
+            ServiceInstance selected = selectInstance(service);
+            assertThat(asString(selected)).isNotEqualTo(asString(last));
+            last = selected;
+        }
+
+        Random random = new Random();
+
+        last = service1;
+        // Start reporting
+        for (int i = 0; i < 102; i++) {
+            if (random.nextInt(10) > 7) {
+                // Simulate failures
+                last.recordResult(1000, new Exception("boom"));
+            } else {
+                last.recordResult(1000, null);
+            }
+
+            ServiceInstance selected = selectInstance(service);
+            assertThat(asString(selected)).isEqualTo(asString(last));
+            last = selected == service1 ? service2 : service1;
+        }
+    }
+
+    @Test
+    void shouldPickBothService() {
+        Service service = stork.getService("first-service");
+
+        Set<String> instances = new HashSet<>();
+
+        for (int i = 0; i < 100; i++) {
+            instances.add(asString(selectInstance(service)));
+        }
+
+        assertThat(instances).hasSize(2).contains(FST_SRVC_1, FST_SRVC_2);
+    }
+
+    @Test
+    void shouldThrowNoServiceInstanceOnNoInstances() throws ExecutionException, InterruptedException {
+        Service service = stork.getService("without-instances");
+
+        CompletableFuture<Throwable> result = new CompletableFuture<>();
+
+        service.selectServiceInstance().subscribe().with(v -> log.error("Unexpected successful result: {}", v),
+                result::complete);
+
+        await().atMost(Duration.ofSeconds(10)).until(result::isDone);
+        assertThat(result.get()).isInstanceOf(NoServiceInstanceFoundException.class);
+    }
+
+    @Test
+    void shouldAlwaysPickFirstWhenTheCallReturnsImmediately() {
+        Service service = stork.getService("first-service");
+
+        Set<String> instances = new HashSet<>();
```

Comment:
note: having a `Set<ServiceInstance>` should be enough

Reply:
No, because we return new instances of `ServiceInstanceWithStatGathering`. It's actually something that puzzled me.